### ISSUE TITLE
Performance discrepancy between `QuantileDMatrix` and `DMatrix`

### ISSUE THREAD

**Issue by dthiagarajan** [dmlc/xgboost#9571]
Hi XGBoost community, 

What kind of model performance discrepancy should I expect when training an XGBoost model using `QuantileDMatrix` vs. `DMatrix`? For reference:
* I am using version 1.7.4 of XGBoost
* I am training trees of max depth 14
* I train distributed, across anywhere between 25-200 workers.

In particular, I'd like to understand what would be necessary to get train the exact same model regardless of whether I use `QuantileDMatrix` vs. `DMatrix`, up to floating point precision.

Thanks in advance!

**Comment by trivialfis** [comment]
> What kind of model performance discrepancy should I expect

Following cases:
- Use in-memory inputs like `numpy.ndarray` or `cudf.DataFrame`, `QDM` should have faster initialization time and lower memory consumption. Part of the initialization of `DMatrix` is deferred to first training iteration so you might find `QDM` to be slower at construction, which is not true.
- Use data iter with batch loading. When `DMatrix` is used, external memory will be enabled and you will be asked for an explicit input for the `prefix` parameter in the data iter init method. This is the slowest option, especially with pre-2.0 versions. XGBoost will load batches during training. With `QDM`, the construction performance depends solely on how fast you can feed the batches into the constructor. The batch loading is only invoked during construction, data is concatenated and compressed during construction. Once construction is finished, the training is the same as in-memory inputs.

> In particular, I'd like to understand what would be necessary to get train the exact same model 

The training algorithm is the same for all kinds of input, they are just different ways for XGBoost to store the intermediate data. `QDM` is an optimization for `hist` to bypass `DMatrix`. Even if you use `DMatrix`, there's a `QDM` generated internally.

**Comment by dthiagarajan** [comment]
Got it, thanks for the thorough answer! So to summarize, if I am using in-memory inputs, I should expect the only discrepancy between `QDM` and `DMatrix` to be due to how the data is stored intermediately? i.e. how it's compressed?

**Comment by dthiagarajan** [comment]
Interesting.. so if I train two models A and B on dataset X, where
* A is trained on QDM(X)
* B is trained on DM(X)

then A and B should be identical models?

**Comment by trivialfis** [comment]
Yup, up to floating-point error, of course. See https://github.com/dmlc/xgboost/blob/f90d034a86784f4f07417d1d28a77b4a189acc89/tests/python/test_quantile_dmatrix.py#L130 .

**Comment by dthiagarajan** [comment]
Got it, thanks! And this would be true even if I train distributed? e.g. if I train on n nodes for each of those cases?

**Comment by dthiagarajan** [comment]
Great, thank you for clarifying!
